### PR TITLE
Tour Chart: Ability to show/hide the following data: Elevation, Distance, Duration, Elevation Gain Difference, Distance Difference, Duration Difference

### PR DESCRIPTION
**Pull request by FJBDev**
image

image

### CODE REVIEW

**Comment by wolfgang-ch** [comment]
> image

I think the tooltip data fields should be displayed in 2 columns to keep this slideout in the width as small as possible and translations may need more width. When adding an additional field, then the width would also grow.

**Comment by FJBDev** [comment]
> I think the tooltip data fields should be displayed in 2 columns to keep this slideout in the width as small as possible and translations may need more width. When adding an additional field, then the width would also grow.

I tried at first but it seems to me that there was enough room for a third column.
But if you prefer 2 columns, I just fixed it:

image

**Comment by wolfgang-ch** [comment]
> > I think the tooltip data fields should be displayed in 2 columns to keep this slideout in the width as small as possible and translations may need more width. When adding an additional field, then the width would also grow.
> 
> I tried at first but it seemed to me that there was enough room for a third column.
> But if you prefer 2 columns, I just fixed it:
> 
> image

With Linux there is a lot of empty space, mainly the spinner control width is about twice as on windows, therefore on Linux it do not look as good as on windows.

The controls could be better placed, delta distance is at the bottom left and distance is at the top right. I have put it in this order

elevation - elevation delta
distance - distance delta
time - time delta

Also the tooltip UI should have this layout without repeating the value name, just showing the delta char.

Is there a need that the user can select the displayed values or just show also the delta values?
Are you planning to show additional values?

**Comment by FJBDev** [comment]
> The controls could be better placed, delta distance is at the bottom left and distance is at the top right. I have put it in this order
> 
> elevation - elevation delta
> distance - distance delta
> time - time delta

Fixed
image


> Also the tooltip UI should have this layout without repeating the value name, just showing the delta char.
> 
> Is there a need that the user can select the displayed values or just show also the delta values?
> Are you planning to show additional values?

**Comment by FJBDev** [comment]
> With Linux there is a lot of empty space, mainly the spinner control width is about twice as on windows, therefore on Linux it do not look as good as on windows.

Also, on Linux the label is not very readable as it is black and the text is grey
image

**Comment by wolfgang-ch** [comment]
> 
> 
> > With Linux there is a lot of empty space, mainly the spinner control width is about twice as on windows, therefore on Linux it do not look as good as on windows.
> 
> Also, on Linux the label is not very readable as it is black and the text is grey
> image

fixed in https://github.com/wolfgang-ch/mytourbook/commit/a5128329f3364d4e9b2ed89e49ba22a2514eae88 the foreground color was not set, this was not an issue when using win 

mt-tour-chart-marker-tooltip

**Comment by wolfgang-ch** [comment]
> 
> 
> I would have tested but my oldest MT backup is from 19.10 (> 18.3)

The field renaming happend in `// 41 -> 42    20.11.1`

**Comment by FJBDev** [comment]
> > I would have tested but my oldest MT backup is from 19.10 (> 18.3)
> 
> The field renaming happend in `// 41 -> 42 20.11.1`

yes but didn't you tell that the database issue only happened with database versions < 18.13 ? `"I wanted to run MT with data before 18.13.1 but it failed"`

**Comment by FJBDev** [comment]
> > > With Linux there is a lot of empty space, mainly the spinner control width is about twice as on windows, therefore on Linux it do not look as good as on windows.
> > 
> > 
> > Also, on Linux the label is not very readable as it is black and the text is grey
> > image
> 
> fixed in [a512832](https://github.com/wolfgang-ch/mytourbook/commit/a5128329f3364d4e9b2ed89e49ba22a2514eae88) the foreground color was not set, this was not an issue when using win10 with default theme
> 
> mt-tour-chart-marker-tooltip

That's looking good on my system!

**Review by wolfgang-ch — COMMENTED** [review]
Because of some text keys which I would have named the same way as other text keys, I've documented the structure of text keys which are currently applied https://mytourbook.sourceforge.io/mytourbook/index.php/development/dev-code-style/i18-code-style. 

I also did some layout/code adjustments in the merged main branch that I don't have to review this again, except the review issues which are not system relevant.

**Comment by wolfgang-ch** [comment]
> I also did some layout/code adjustments in the merged main branch

Before

mt-tooltip-before


After

mt-tooltip-after

File: bundles/net.tourbook.common/src/net/tourbook/common/util/StringUtils.java [hunk 1]
```diff
@@ -61,4 +63,13 @@ public static String join(final String[] stringArray, final String separator) {
...
       return sb.toString();
    }
+
+   public static String sanitizeFileName(final String fileName) {
+
+      if (StringUtils.isNullOrEmpty(fileName)) {
+         return fileName;
+      }
+
+      return fileName.replaceAll("[^a-zA-Z0-9 \\.\\-]", UI.DASH); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.common/src/net/tourbook/common/util/StringUtils.java:73]: You mentioned it [here](https://github.com/wolfgang-ch/mytourbook/pull/299#discussion_r571380786) and it is fixed in this [PR](https://github.com/wolfgang-ch/mytourbook/pull/301) with other fixes

File: bundles/net.tourbook/src/net/tourbook/preferences/ITourbookPreferences.java [hunk 2]
```diff
@@ -139,58 +139,72 @@
...
+   public static final String GRAPH_MARKER_IS_MODIFIED                                     = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
+   public static final String GRAPH_MARKER_IS_LABEL_ELEVATION                              = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
+   public static final String GRAPH_MARKER_IS_LABEL_DISTANCE                               = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
+   public static final String GRAPH_MARKER_IS_LABEL_TIME                                   = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
+   public static final String GRAPH_MARKER_IS_LABEL_TIME_DIFFERENCE                        = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
+   public static final String GRAPH_MARKER_IS_LABEL_DISTANCE_DIFFERENCE                    = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
+   public static final String GRAPH_MARKER_IS_LABEL_ELEVATIONGAIN_DIFFERENCE               = "GRAPH_MARKER_IS_MODIFIED";                               //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/preferences/ITourbookPreferences.java:170]: Fixed in this [PR](https://github.com/wolfgang-ch/mytourbook/pull/301)

File: bundles/net.tourbook/src/net/tourbook/ui/views/rawData/RawDataView.java [hunk 3]
```diff
@@ -448,26 +450,28 @@
...
-   private Link      _linkImport;
+   private Link                          _linkImport;
 
-   private Browser   _browser;
+   private Browser                       _browser;
 
-   private Menu      _tableContextMenu;
+   private Menu                          _tableContextMenu;
+
+   private List<TourbookCloudDownloader> _cloudDownloadersList = CloudDownloaderManager.getCloudDownloaderList();
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/ui/views/rawData/RawDataView.java:474]: Fixed in this [PR](https://github.com/wolfgang-ch/mytourbook/pull/301)

File: bundles/net.tourbook/src/net/tourbook/ui/views/rawData/RawDataView.java [hunk 4]
```diff
@@ -4285,6 +4297,11 @@ private void onBrowser_LocationChanging_Runnable(final String[] locationParts) {
       } else if (ACTION_OLD_UI.equals(hrefAction)) {
 
          onSelectUI_Old();
+      } else {
+
+         _cloudDownloadersList.stream()
+               .filter(cd -> cd.getId().equals(hrefAction))
+               .forEach(TourbookCloudDownloader::downloadTours);
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/ui/views/rawData/RawDataView.java:4304]: Fixed in this [PR](https://github.com/wolfgang-ch/mytourbook/pull/301)

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/suunto/SuuntoCloudDownloader.java [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+         TourLogManager.logTitle(Messages.Log_DownloadWorkoutsToSuunto_001_Start);
+
+         new ProgressMonitorDialog(Display.getCurrent().getActiveShell()).run(true, false, runnable);
+
+         TourLogManager.logTitle(String.format(LOG_CLOUDACTION_END, (System.currentTimeMillis() - start) / 1000.0));
+
+         MessageDialog.openInformation(
+               Display.getDefault().getActiveShell(),
+               Messages.Dialog_WorkoutsDownload_Summary,
+               NLS.bind(Messages.Dialog_WorkoutsDownload_Message,
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/suunto/SuuntoCloudDownloader.java:218]: Fixed in this [PR](https://github.com/wolfgang-ch/mytourbook/pull/301)

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/suunto/Messages.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+   public static String        Pref_AccountInformation_SuuntoApp_WebPage_Link;
+   public static String        Pref_Checkbox_Use_SinceDateFilter;
+   public static String        Pref_Checkbox_Use_SinceDateFilter_Tooltip;
+   public static String        Pref_Combo_Workouts_FolderPath_Combo_Tooltip;
+
+   public static String        Suunto_Workouts_Description;
+   public static String        UploadToursToSuunto_Task;
+   public static String        UploadToursToSuunto_SubTask;
+   public static String        VendorName_Suunto_Routes;
+   public static String        VendorName_Suunto_Workouts;
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/suunto/Messages.java:54]: After [this comment](https://github.com/wolfgang-ch/mytourbook/pull/299#discussion_r571571580) last week, this file doesn't exist anymore and the grouping should be better. See this [PR](https://github.com/wolfgang-ch/mytourbook/pull/301).

image

**Comment by FJBDev** [comment]
> 
> Because of some text keys which I would have named the same way as other text keys, I've documented the structure of text keys which are currently applied https://mytourbook.sourceforge.io/mytourbook/index.php/development/dev-code-style/i18-code-style.
> 
Ok, I will keep that documentation handy.
What does i18 mean ?

> I also did some layout/code adjustments in the merged main branch that I don't have to review this again, except the review issues which are not system relevant.
Sure. Sometimes it's better that you make the modifications so you can make the exact modifications you have in mind. That looks good to me.

**Comment by wolfgang-ch** [comment]
> What does i18 mean ?

I removed the n from i18n (it looks better :-) 

i18n: https://en.wikipedia.org/wiki/Internationalization_and_localization

**Comment by FJBDev** [comment]
> 
> 
> > What does i18 mean ?
> 
> I removed the n from i18n (it looks better :-)
> 
> i18n: https://en.wikipedia.org/wiki/Internationalization_and_localization

Ok, I didn't know that.

**Pull request merged**